### ISSUE TITLE
[RayService] Ray cluster keeps restarting when autoscaling with RayServices

### ISSUE THREAD

**Issue by shrekris-anyscale** [ray-project/kuberay#649]
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/kuberay/issues) and found no similar issues.

### KubeRay Component

Others

### What happened + What you expected to happen

I deployed a RayService on a Kubernetes cluster with cluster autoscaling enabled via `enableInTreeAutoscaling`. See [this gist](https://gist.github.com/shrekris-anyscale/bdd6607fae33f5303cea9c13b7638d57) for the config file. The cluster comes up successfully, but it's then terminated and a new cluster comes up. This happens repeatedly and doesn't seem to stop.

### Reproduction script

I deployed the config stored in [this gist](https://gist.github.com/shrekris-anyscale/bdd6607fae33f5303cea9c13b7638d57):

```bash
kubectl create -k "github.com/ray-project/kuberay/ray-operator/config/default?ref=master&timeout=90s"
```

```console
$ kubectl get deployments -n ray-system
NAME                READY   UP-TO-DATE   AVAILABLE   AGE
kuberay-operator    1/1     1            1           13s
```

```bash
k apply -f rayservice_autoscale_loop.yaml
```

This should spin up 4 pods (1 head pod and 3 worker pods). The config file sets the minimum number of worker pods to 2, so after some time without sending any requests, the cluster will attempt to scale down. However, it will never actually scale down. Instead, a new cluster will spin up and the old one will be terminated. E.g.:

```console
$ kubectl get pods
NAME                                             READY   STATUS        RESTARTS   AGE
rxam-raycluster-2rzxj-head-pz246                 2/2     Running       0          35s
rxam-raycluster-2rzxj-worker-small-group-95jmg   1/1     Running       0          35s
rxam-raycluster-2rzxj-worker-small-group-l69hg   1/1     Running       0          35s
rxam-raycluster-2rzxj-worker-small-group-xwzz2   1/1     Running       0          35s
rxam-raycluster-kgp8h-head-w787z                 2/2     Terminating   0          2m6s
rxam-raycluster-kgp8h-worker-small-group-9552l   1/1     Terminating   0          2m6s
rxam-raycluster-kgp8h-worker-small-group-jxnl9   1/1     Terminating   0          2m6s
```

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

**Comment by simon-mo** [comment]
The root cause of this that during scaling, RayCluster operator will mutate the spec of the cluster changing worker `Replicas` and `WorkersToDelete` fields. For this given workload, we can see these around 60-90s mark
```
# before in spec.worker section
    Replicas:               3
    Scale Strategy:

# after in spec.worker section
    Replicas:               2
    Scale Strategy:
      Workers To Delete:
        rxam-raycluster-hn69g-worker-small-group-pftww
```

This causes some issue in the RayService operator where it actively check the drift between desired spec and actual spec. If the the desired != actual spec, it will restart the cluster:
https://github.com/ray-project/kuberay/blob/389ba0042728c3611c8596483ff442407fe08360/ray-operator/controllers/ray/rayservice_controller.go#L253-L256

https://github.com/ray-project/kuberay/blob/389ba0042728c3611c8596483ff442407fe08360/ray-operator/controllers/ray/rayservice_controller.go#L325-L337


The logic used for comparison is just comparing the JSON representation of both objects: https://github.com/ray-project/kuberay/blob/389ba0042728c3611c8596483ff442407fe08360/ray-operator/controllers/ray/utils/util.go#L316-L336

This means during scaling (up or down, doesn't matter), RayService operator recognizes the mutated spec (which RayCluster uses to keep track of desired replica) as a drift, therefore creating a new cluster. (logging messages like `Current cluster is unhealthy, prepare to restart...`).

**Comment by simon-mo** [comment]
Now onto the fix:
- hacky fix: the JSON comparison should ignore fields that can be mutated by RayCluster operator (desired worker replica, workers to delete etc).
- slightly better fix: the RayService operator should not restart the cluster when "the active RayCluster has changed." Instead of comparing them 
```golang
utils.CompareJsonStruct(activeRayCluster.Spec, rayServiceInstance.Spec.RayClusterSpec)
```
it should just blindly apply the spec and let RayCluster operator handle it.
- another fix (?): we should distinguish between a real ray cluster update made by human, as compare to the ones made by machines. this means, we should only create new clusters when the new embedded ray cluster configuration is different than the previously submitted version; instead of comparing the drift of desired YAML state vs current state. However, this method is probably not very robust and goes against the [level triggered design principle in Kubernetes](https://speakerdeck.com/thockin/edge-vs-level-triggered-logic)
- proper fix: the proper fix is that NO OPERATOR SHOULD BE ALLOWED TO EDIT THE SPEC SECTION. Instead, it should store all the necessary states it needs in status section.

**Comment by DmitriGekhtman** [comment]
For the immediate moment, I endorse the hacky fix.
Once a correct story has been worked out for RayCluster pod config updates, we can follow the "update the spec and let the RayCluster controller handle it" approach.

**Comment by DmitriGekhtman** [comment]
The autoscaler is not an operator -- from the RayCluster controller's perspective it's equivalent to a human modifying the replicas field.
The replicas field is the interface for any human or machine scaler to do what it wants.

**Comment by simon-mo** [comment]
@DmitriGekhtman the main issue is the workers-to-delete is not mutated by autoscaler, rather by the operator itself, which performs a mutation of the spec section on the reconcile path.

**Comment by simon-mo** [comment]
Now i'm looking at the code, it's going to be difficult to figure out what can RayCluster mutate. I'm leaning towards storing previous rayCluster spec into annotation and perform a comparison. This is how, for example, `kubectl` perform a diff. https://kubernetes.io/docs/tasks/manage-kubernetes-objects/declarative-config/#how-to-create-objects

**Comment by DmitriGekhtman** [comment]
Ah, yeah, I see your point!
We might want to revisit workersToDelete.

Anyways, for now I advocate hack -- ignore scaleStrategy and replicas.

**Comment by DmitriGekhtman** [comment]
> Now i'm looking at the code, it's going to be difficult to figure out what can RayCluster mutate. I'm leaning towards storing previous rayCluster spec into annotation and perform a comparison. This is how, for example, `kubectl` perform a diff. https://kubernetes.io/docs/tasks/manage-kubernetes-objects/declarative-config/#how-to-create-objects

The strategy of storing last applied configuration in the database is definitely more robust than storing in memory. I think we went with the first approach because it was easier to implement for version 0.

**Comment by DmitriGekhtman** [comment]
I think I'm actually thinking about another config comparison in the last comment -- anyways, this problem is fixable.

**Comment by simon-mo** [comment]
@shrekris-anyscale will be working on the fix here. The plan is to store a hash of the cluster spec and only create new clusters when the hash no longer matches. 

In more details
- Step 1: User applies a config
  - The service operator will hash the cluster spec to XYZ. 
  - The service operator will add a metadata label the cluster config saying serve-cluster-config-hash: XYZ. 
  - The service operator will do its normal thing.
…
- In the drift detection workflow:
  - If the hash matches, don’t compare. 
  - If it doesn’t, create new cluster and perform switch when healthy
- Step N: User re-apply with changed cluster config
  - Hash it again. Compare it against live config’s label serve-cluster-config-hash. 
  - See it differs, then create a new pending cluster for the new config. 
  - Do the normal A/B flow. 

This is pretty conservative because any cluster spec change will invoke an A/B update. We think this is fine for now because the risk of directly applying the new config to Ray Cluster operator is quite high as it can 
- Reset desired replica for worker group
- Perform container refreshes with downtime
- Wipe out the `workersToDelete` implementation details, therefore cause potential leaks.

**Comment by DmitriGekhtman** [comment]
Just to be clear here, is this correct:
We are not going to include workersToDelete or replicas in the hash, right?

**Comment by DmitriGekhtman** [comment]
Assuming the last comment, this solution seems ok for now. Longer term, we will need to also have an update strategy for bare RayClusters and make sure that the update strategy for RayServices meshes well.
See https://github.com/ray-project/kuberay/issues/527 and linked issues for the discussion about RayCluster upgrades.

**Comment by shrekris-anyscale** [comment]
I'm reclassifying this issue as P1 since #655 resolves the Ray cluster restart issue. I'll leave this issue open since there's some discussion about long-term updates we can make to further address this issue.

**Comment by joanna-yoo** [comment]
Hi, is this issue fully fixed? I am experiencing similar problems at the moment. I am using config similar to [ray-cluster.autoscaler.large.yaml](https://github.com/ray-project/kuberay/blob/cd9b2e8b2c3598b4f792405c9ceb4b4bcdde2cdd/ray-operator/config/samples/ray-cluster.autoscaler.large.yaml) with ray 2.2.0 and kuberay 0.4.0, but I think I experience the same issue.

The ray head (and the workers) are running in healthy state, but a new ray head is created and kills the existing head and workers.

Is there a way to fix this? 
```
example-raycluster-mr55q-head-dj47m                     2/2     Running             0          22m
example-raycluster-5hs88-head-nzgtb                     0/2     Pending             0          0s
```


```
example-raycluster-mr55q-head-dj47m                     2/2     Terminating         0          24m
example-raycluster-5hs88-head-nzgtb                     2/2     Running             0          2m24s
```

**Comment by kevin85421** [comment]
Hi @joanna-yoo, could you provide your YAML file? Is it similar to https://github.com/ray-project/kuberay/issues/1030? Thanks!

**Comment by kevin85421** [comment]
This issue has already been resolved by #655.

```sh
kind create cluster --image=kindest/node:v1.23.0
helm install kuberay-operator kuberay/kuberay-operator --version 0.5.0
# https://gist.github.com/shrekris-anyscale/bdd6607fae33f5303cea9c13b7638d57
kubectl apply -f test.yaml

# The RayCluster can successfully scale down without the need to prepare a new RayCluster.
```

In addition, the other parts related to autoscaler support have been resolved by #1037. Close this issue. Feel free to reopen this issue if necessary.
